### ISSUE TITLE
Why visibility and const

### ISSUE THREAD

**Issue by 0dminnimda** [robertmuth/Cwerg#41]
Most of the features listed in the `tutorial.md` are useful and are improvements over C. Good work, keep it up! But I genuinely don't understand why:

## Add visibility at all

I really like the idea of "we are all consenting adults here." If I want to make a custom function for your data structure, let me hack at it; I take responsibility! I can pay for maintaining the state with improved performance for my specific case, or a nicer calling convention, or... And is it a systems programming language or not? Why is my control reduced?

## Add const/mutability correctness

It's really not nice to have this spreading like async; [colored functions](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) are not fun.

It is really useful to have global constants and even local ones, but adding constness to the function args or every variable is a bit much. Since conditional logic requires explicit bool, the `if (a = 5)` typo is no longer a problem. I don't really think that unintended modification of the value passed by reference is really a big problem, and it's more of a logical error either way. All the optimization we could get is covered by global constants or in C `#define CONSTANT 5`. So the only use I see is compiler-enforced documentation. Here I'm struggling; I cannot recall having a problem of unintended modification. What was your motivation to include this feature? What actual problem does const correctness solve? For what is the language user paying?

I'm assuming special syntax for constants and no special syntax for regular variables.

**Comment by robertmuth** [comment]
I have worked on some really really large c++ code bases and that  experience has informed the design choices that  Cwerg makes.
Cwerg optimizes for code reading rather than code writing.
So compiler enforced documentation is the goal and colored functions are a good trade-off  
at least in the case of const correctness.

RE: Add visibility at all

I find it extremely helpful to know that certain "state" can only be accesed and/or changed by
code in the  "vicinity". It makes reasoning about a large program possible and simplifies refactoring.
If you need to work around visibility you have several options. e.g.:
* add code the module that has visibility
* use (unsafe/creative) pointer casting  

RE: const/mutability correctness

Compare:
1) foo(T* p) 
2) foo(const T* p)

For 2) a caller can assume that  foo will not change *p  for 1) it cannot.
When you work in large code bases this is invaluable.
It also helps with optimization.

When it comes to variables:
let x = 5   (immutable case) puts a lot less cognitive load  on the reader than "let! x = 5" (mutable case).
E.g.: if a piece of code updates 10 variables in its scope - it is borderline spaghetti. 10 constants declared in that  scope are not a big deal.

Basically, mutable state is bad. But if you have to use it, Cwerg will make you work a little harder.

**Comment by 0dminnimda** [comment]
Ok, visibility is just a preference. No bike shedding.

---

Obviously, compiler-enforced documentation is much better than non-enforced one. This won't decay nearly as quickly.

But about foo(T* p).

1. It would be useful to separate the "out" pointer from the "update"/"read" pointer. Since the "out" should be handled by the ability to return a tuple/multi-variable return. This way the intent is much clearer.

2. Now I too definitely do feel better when I add `const` to parameters that I don't change in C, but I usually don't even think about whether something has been changed or not when I do the call in Python, almost like with the right design it doesn't matter that much 🤔. Is it just a generalization of 'reduce mutable state' with 'it feels better'? Maybe you have a more concrete example? I feel that the difference between "update" and "read" pointers is usually not important to the caller; the outer logic is the same.

> mutable state is bad

So an actually useful version of FP might be to reduce mutable state to a minimum

> It also helps with optimization.

I can only think of the compiler taking advantage of multiple cores, doing some passes in parallel and not doing some proofs, but not the actual optimization of the produced code. I am probably ignorant here, but since you'll have to convert to SSA, you'll know what is updated and at what places, and then it's really easy to constantly propagate even values not marked as a constant.

**Comment by robertmuth** [comment]
Keep in mind that Cwerg is trying to be small. It is close to hitting the 10k LOC limit and I'd like to keep some LOC budget around for optimizations. 
Multiple return values are a nice-to-have and the backend even supports them but I could not make it work under the given complexity constraints in the front end.
Marking parameters as in, out, in/out would also nice but  my guesstimate  is that in/out dominates out by a wide margin.

> Now I too definitely do feel better when I add const to parameters that I don't change in C, but I usually don't even think about whether something has been changed or not when I do the call in Python, almost like with the right design it doesn't matter that much 🤔. Is it just a generalization of 'reduce mutable state' with 'it feels better'? Maybe you have a more concrete example? I feel that the difference between "update" and "read" pointers is usually not important to the caller; the outer logic is the same.

Once the code base is large enough and you cannot keep the whole project in your head anymore, 
it becomes really hard to figure out who is modifying what. In C++ you can make this extra hard by using either non-const 
pointers or non-const references.  I have not looked into it in a while but LLVM used to have this problem/anti-pattern.

> It also helps with optimization.

I am only starting with implementing optimizations in Cwerg.  One place where I found "const" useful: Cwerg does generate a lot of copies, e.g. after inlining and for various desugaring phases.  If you have  a statement like "let a = b" AND you know that b is  const you can just eliminate a.  Can this information be recovered through analysis? Probably. But why not use the info the programmer provided and safe some LOC budget.
BTW:  The Cwerg BE currently does not use SSA. And I'd like to do these optimizations in the FE anyway.

**Comment by 0dminnimda** [comment]
> I could not make it work under the given complexity constraints in the front end

Fair enough!

> Once the code base is large enough and you cannot keep the whole project in your head anymore,
It becomes really hard to figure out who is modifying what. 

True, but why does it matter who is modifying what? Isn't the point of an abstraction to allow you to forget about the specific implementation? And also, shouldn't you try to separate code into separate atomic parts that don't do crazy changes to all reachable data?

Of course we don't run our programs on magical machines; you have to take into consideration the details, but up to a certain point. Sometimes the optimization does not matter there, sometimes it's so grand you cannot possibly have a map of the code, then you operate on abstractions. And at that level the micro-optimizations usually don't matter much.

> Can this information be recovered through analysis? Probably. But why not use the info the programmer provided and safe some LOC budget.

Good point. I'm fond of the idea of allowing programmers to better express the intent, thus simplifying the compiler, allowing it to focus on more important tasks, and making it simpler to apply optimization passes.

Though it should not come at a cost for the routine and most frequent tasks.

>  The Cwerg BE currently does not use SSA

You have got to read this paper [Simple and Efficient Construction of Static Single Assignment Form.pdf](https://github.com/user-attachments/files/18141213/Simple.and.Efficient.Construction.of.Static.Single.Assignment.Form.pdf)! Compact implementation for SSA with code example. From the personal experience I'd argue that the code will become simpler and more open to optimizations with SSA, you don't need to have separate 'temps' and 'variables' and a lot of things just work better.

**Comment by robertmuth** [comment]
> True, but why does it matter who is modifying what? Isn't the point of an abstraction to allow you to forget about the specific implementation? And also, shouldn't you try to separate code into separate atomic parts that don't do crazy changes to all reachable data?

This happens all the time, especially in compiler writing: The compiler miscompiles. Some invariant was violated or some analysis state got accidentaly invalidated and now the chase for the culprit starts. While reading through the source code ,
const pointer like things allow you to quickly rule out whole subtrees of code - assuming that the pointer is the only way for the subtree to access that data, i.e. no global contain a non-const version of the pointer.
Moving away from raw data: in C++ you may have member functions that update state and those that only read state, if you pass a const pointer to such an object, you know that no updates can happens. 
Being very disciplined can help a lot, but the larger and older the project gets, the more likely it become that you run into problems like the one I sketched above.

> You have got to read this paper [Simple and Efficient Construction of Static Single Assignment Form.pdf](https://github.com/user-attachments/files/18141213/Simple.and.Efficient.Construction.of.Static.Single.Assignment.Form.pdf)!

Thanks for reminding me about SSA . For my own benefit, I also note: 
https://www.reddit.com/r/Compilers/comments/1hdhx8s/lowering_of_phi_nodes/

I will revisit SSA once the frontend has been re-implmented in  C++.
